### ISSUE TITLE
Rita.untokenize incorrectly handling apostrophe's at end of word

### ISSUE THREAD

**Issue by dhowe** [dhowe/RiTa#477]

**Comment by dhowe** [comment]
Here is problematic case in our untokenize tests:

a. The student said 'learning is fun'
b. We should consider the students' learning

In each case, when we get to the single-quote, we don't know whether it is the start of a new quotation or an apostrophe at the end of the previous word

@cqx931 any thoughts?

**Comment by cqx931** [comment]
When I test RiTa.tokenize(), `student’s` stay as one token while 'students'' become two. 
But if  `student’s` is regarded as one token, shouldn't 'students'' also stay as one? And in this case it would be an issue for tokenize(), rather than untokenize().

If we do think 'students'' should be broke into two tokens, then are words ending in 's' the only case here? (I can't come up with more cases... ) If it is the only case we could check whether the token before is ending in 's' to differentiate case a and b.

**Comment by dhowe** [comment]
This is a good question. I think punctuation that is internal to a word (e.g., student's) should be a single token, while punctuation at end of word (e.g., students' or students" or students, or students:) should be two. One reason is that we don't want the 's' character in "student's" to be a separate token...

**Comment by dhowe** [comment]
In any case, all tests are passing now except for the following, which I've added to knownIssues-tests:

      var expected = "We should consider the students' learning.";
      var input = ["We", "should", "consider", "the", "students", "'", "learning" ];
      var output = RiTa.untokenize(input);
      deepEqual(output, expected);

@cqx931 please update RiTa (including the tests) to match the behavior of RiTaJS... you may want to simply rewrite untokenize() as ported from JS

**Comment by cqx931** [comment]
First fixed the issue in tokenize():
https://github.com/dhowe/RiTaJS/pull/88/commits/cfa2e90eb856a3d946b29e0f771f7a925d538b07

**Comment by cqx931** [comment]
Please check: https://github.com/dhowe/RiTaJS/pull/89
Fixed the rest issues with untokenize() and added other curly quotes to tokenize().

**Comment by dhowe** [comment]
Excellent. Now please 1) update the RiTa java tests to match those in JS, then 2) update RiTa.tokenize/untokenize() functions in java to match the behavior of RiTaJS... you may want to simply rewrite those functions as ports from JS, rather than trying to adapt what's there now

**Comment by cqx931** [comment]
Please check https://github.com/dhowe/RiTa/pull/480
and https://github.com/dhowe/RiTaJS/pull/90

**Comment by cqx931** [comment]
Can this be closed?

**Comment by dhowe** [comment]
Do we have this case in KnownISsues for both platforms?

 ```
     var expected = "The student said 'learning is fun'";
      var input = ["The", "student", "said", "'", "learning", "is", "fun", "'"];
      var output = RiTa.untokenize(input);
      deepEqual(output, expected);
```

**Comment by cqx931** [comment]
This test case is already solved for both platforms (in RiTaTest).

**Comment by dhowe** [comment]
So we have both of these sentences in the tests:

**Comment by dhowe** [comment]
So we have both of these arrays as inputs to untokenize tests:


1.  ["We", "should", "consider", "the", "students", "'", "learning" ]
1.  ["The", "student", "said", "'", "learning", "is", "fun", "'"]

Can you explain how the algorithm groups the first single-quote in 1 with the previous word ('students''), but  to do the opposite in 2, and group it with the subsequent word (''learning')?

**Comment by dhowe** [comment]
@cqx931 ?

**Comment by cqx931** [comment]
What do you mean by grouping here? Isn't both in case 1 and case 2, that the single quote is considered as a single token?

**Comment by dhowe** [comment]
Yes, but in the output to untokenize... (it outputs a single string)

**Comment by dhowe** [comment]
@kennyviperhk Please verify that both of the tests below pass (in both RiTa and RiTaJS) for RiTa.untokenize:

```
    expected = "We should consider the students' learning";
    input = ["We", "should", "consider", "the", "students", "'", "learning" ]
    var output = RiTa.untokenize(input);
    equal(output, expected);

    expected = "The student said 'learning is fun'";
    input = ["The", "student", "said", "'", "learning", "is", "fun", "'"]
    var output = RiTa.untokenize(input);
    equal(output, expected);
```
